Add a versatile nodeset generator

GiudGiud opened the pull request:
closes https://github.com/idaholab/moose/issues/5895
refs https://github.com/idaholab/moose/issues/28881

To get around difficulties reported in #28869 using the Gmsh reader.
This is not the best motivation, but I can see this being useful regardless

GiudGiud committed:
Clean up ParsedGenerateSideset

GiudGiud committed:
Add a base class for nodeset generation with a few of the commonly desirable options
- included/excluded boundaries
- included subdomains
- external only
refs #28881

GiudGiud committed:
Add a nodeset generator based on a parsed expression
refs #28881

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2502285/), step Docs: sync website on 2824a41 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/28873/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2502300/), step Generate coverage on 2824a41 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>b3924e</th>
      <th colspan=4>#28873 2824a4</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.05%</td>
      <td>84.96%</td>
      <td>-0.08%</td>
      <td>15.72%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>106298</td>
      <td>106322</td>
      <td>+24</td>
      <td>25</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18691</td>
      <td>18820</td>
      <td>+129</td>
      <td>134</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28873/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28873/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/28873/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/28873/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/28873/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/28873/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/28873/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/28873/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/28873/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/28873/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/28873/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/28873/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/28873/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/28873/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/28873/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/28873/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/28873/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/28873/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/28873/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/28873/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/28873/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/28873/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/28873/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/28873/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/28873/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/28873/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/28873/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/28873/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/28873/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/28873/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 15.72% is less than the suggested 90.0%



This comment will be updated on new commits.

GiudGiud committed:
Add tests for parsed nodeset generator:
- exceptions wrong parameter
- exception missing bdy / subdomain (#22117)
- regular test for each feature

GiudGiud commented:
@pbehne would you have time to review? This is very close to what you did for sidesets

pbehne commented:
> @pbehne would you have time to review? This is very close to what you did for sidesets

I'll take a look in the coming week!

pbehne commented:
Starting on this now.

GiudGiud committed:
Address Patrick's review

- fix docstrings
- add more param details in docstring
- more const

Co-authored-by: Patrick Behne <31858053+pbehne@users.noreply.github.com>

GiudGiud commented on framework/include/meshgenerators/NodeSetsGeneratorBase.h:
```diff
@@ -0,0 +1,100 @@
+   */
+  void setup(MeshBase & mesh);
+
+  /**
+   * Determines whether the node is on the exterior of the mesh
+   */
+  bool nodeOnMeshExteriorBoundary(const Node * node,
+                                  const std::vector<dof_id_type> & node_elems,
+                                  const MeshBase & mesh) const;
```
done

GiudGiud commented on framework/include/meshgenerators/NodeSetsGeneratorBase.h:
```diff
@@ -0,0 +1,100 @@
+  /**
+   * Determines whether the given element's node satisfies the user-specified constraints
+   */
+  bool nodeSatisfiesRequirements(const Node * node,
```
for sure

GiudGiud commented on framework/src/meshgenerators/ParsedGenerateNodeset.C:
```diff
@@ -0,0 +1,142 @@
+  setup(*mesh);
+
+  if (!mesh->is_replicated())
+    mooseError("Not implemented for distributed meshes");
```
not tested in any MG afaik. Here we are hopeful someone will come in and implement it

GiudGiud commented on framework/src/meshgenerators/ParsedGenerateSideset.C:
```diff
@@ -70,12 +68,7 @@ ParsedGenerateSideset::ParsedGenerateSideset(const InputParameters & parameters)
-               name(),
-               ".\n",
-               _func_F->ErrorMsg());
+    mooseError("Invalid function\n", _function, "\n", _func_F->ErrorMsg());
```
made consistent

GiudGiud replied:
we could add a test. maybe in a dedicated PR on catching all MG exceptions missing coverage

GiudGiud commented on framework/src/meshgenerators/ParsedGenerateNodeset.C:
```diff
@@ -0,0 +1,142 @@
+    _func_params[2] = (*curr_node)(2);
+    if (evaluate(_func_F))
+    {
+      if (_replace)
```
done

GiudGiud commented on framework/src/meshgenerators/ParsedGenerateNodeset.C:
```diff
@@ -0,0 +1,142 @@
+               "\nin ParsedGenerateNodeset ",
+               name(),
+               ".\n",
+               _func_F->ErrorMsg());
```
added

GiudGiud commented on framework/include/meshgenerators/NodeSetsGeneratorBase.h:
```diff
@@ -0,0 +1,100 @@
+   * Determines whether the given node of an element belongs to any nodesets in the
+   * included_nodesets parameter.
+   */
+  bool nodeInIncludedNodesets(const std::vector<BoundaryID> & node_nodesets) const;
```
done

GiudGiud commented:
thanks!
@roystgnr @lindsayad for the ccb review?

GiudGiud commented:
or @cticenhour if you want newsletter content 📰

lindsayad commented on framework/include/meshgenerators/NodeSetsGeneratorBase.h:
```diff
@@ -0,0 +1,111 @@
+  /// A map from nodes (ids) to  local elements (ids) which comprise the node
+  // Build the node to element map, which is usually provided by a MooseMesh but here we have a
+  // MeshBase.
+  std::map<dof_id_type, std::vector<dof_id_type>> _node_to_elem_map;
```
```suggestion
  std::unordered_map<dof_id_type, std::vector<dof_id_type>> _node_to_elem_map;
```

GiudGiud committed:
Address Alex's review

- use an unordered map for performance

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

GiudGiud commented:
Test failures are unrelated.

Thanks for the reviews!
